### ISSUE TITLE
Problem in loss normalization

### ISSUE THREAD

**Issue by atebbifakhr** [OpenNMT/OpenNMT-tf#570]
Hi,

In `cross_entropy_sequence_loss` function, if `average_in_time` is `True`, the loss computed for each sequence should be divided by the length of the same sequence. But it's divided by the sum of the length of all the sequences in the batch.

**Comment by guillaumekln** [comment]
This is the expected behavior as `cross_entropy_sequence_loss` always averages the loss across the batch dimension. `average_in_time` is to also include the time dimension.

**Comment by atebbifakhr** [comment]
I agree with the first sentence. But when the `average_in_time` is `True` instead of:
```
loss_num = tf.reduce_sum(cross_entropy * weights)
loss_den = tf.reduce_sum(weights)
loss = loss_num / loss_den
```
I think it should be:
```
loss_num = tf.reduce_sum(cross_entropy * weights, axis=1)
loss_den = tf.reduce_sum(weights, axis=1)
loss = tf.reduce_sum(loss_num/loss_den)
```

**Comment by guillaumekln** [comment]
The argument name is possibly badly chosen but when `average_in_time` is `True` in practice we want to divide the loss by the number of tokens in the batch.

You can see it like this (for fixed-length sequences):

```python
loss_num = tf.reduce_sum(cross_entropy)
loss_den = 1
if average_batch:  # assumed to be True in the code.
  loss_den *= tf.shape(cross_entropy)[0]
if average_in_time:
  loss_den *= tf.shape(cross_entropy)[1]
loss = loss_num / loss_den
```

**Comment by atebbifakhr** [comment]
I thought `average_in_time` was supposed to normalize the effect of length variability. In this way, longer sequences in the batch contribute more in the final loss.

**Comment by guillaumekln** [comment]
> In this way, longer sequences in the batch contribute more in the final loss.

This is what happens when `average_in_time` is `False`, i.e. longer sentences will produce a higher loss. Am I missing something?

**Comment by atebbifakhr** [comment]
But it also happens when `average_in_time` is `True`.
Basically longer sequences have higher loss, since it is sum of computed losses for each token. In order to normalize this phenomena, the loss of each sequence should be divided by its own length.

**Comment by guillaumekln** [comment]
Do you have experiments with that? As far as I know, most NMT implementations sum the loss and divide by the number of tokens. For RNN, we used to divide by the batch size instead.
